File: email_extractor/core_php/storage.php
<?php
function init_db(string $dbFile = 'emails.db'): SQLite3 {
    $db = new SQLite3($dbFile);
    $db->exec('CREATE TABLE IF NOT EXISTS emails (email TEXT PRIMARY KEY, title TEXT, website TEXT)');
    return $db;
}
function save_email(string $email, string $title = '', string $website = '', string $dbFile = 'emails.db'): void {
    $db = init_db($dbFile);
    $stmt = $db->prepare('INSERT OR IGNORE INTO emails(email,title,website) VALUES (:e,:t,:w)');
    $stmt->bindValue(':e',$email,SQLITE3_TEXT); $stmt->bindValue(':t',$title,SQLITE3_TEXT); $stmt->bindValue(':w',$website,SQLITE3_TEXT); $stmt->execute(); $db->close();
}
function export_csv(string $filename = 'emails.csv', string $dbFile = 'emails.db'): void {
    $db = init_db($dbFile); $res = $db->query('SELECT email,title,website FROM emails');
    $f = fopen($filename, 'w'); fputcsv($f, ['Email','Title','Website']);
    while ($row = $res->fetchArray(SQLITE3_ASSOC)) fputcsv($f, [$row['email'],$row['title'],$row['website']]);
    fclose($f); $db->close();
}

File: email_extractor/core_rust/storage.rs
use rusqlite::{params, Connection};
use std::fs::File;
use std::io::Write;

pub fn init_db(db: &str) -> rusqlite::Result<()> {
    let conn = Connection::open(db)?;
    conn.execute("CREATE TABLE IF NOT EXISTS emails (email TEXT PRIMARY KEY, title TEXT, website TEXT)", [])?;
    Ok(())
}
pub fn save_email(db: &str, email: &str, title: &str, website: &str) -> rusqlite::Result<()> {
    let conn = Connection::open(db)?;
    conn.execute("INSERT OR IGNORE INTO emails(email,title,website) VALUES (?1,?2,?3)", params![email,title,website])?;
    Ok(())
}
fn csv(s: &str) -> String { format!("\"{}\"", s.replace('"', "\"\"")) }
pub fn export_csv(db: &str, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    init_db(db)?; let conn = Connection::open(db)?;
    let mut stmt = conn.prepare("SELECT email,title,website FROM emails")?;
    let rows = stmt.query_map([], |r| Ok((r.get::<_,String>(0)?, r.get::<_,String>(1)?, r.get::<_,String>(2)?)))?;
    let mut file = File::create(filename)?; writeln!(file, "Email,Title,Website")?;
    for row in rows { let (e,t,w) = row?; writeln!(file, "{},{},{}", csv(&e),csv(&t),csv(&w))?; }
    Ok(())
}
